src/python/base/utils.py

```diff
@@ -782,16 +787,20 @@ def wait_until_timeout(threads, thread_timeout):
 
 def write_data_to_file(content, file_path, append=False):
   """Writes data to file."""
-  content_string = str(content)
   failure_wait_interval = environment.get_value('FAIL_WAIT')
   file_mode = 'ab' if append else 'wb'
   retry_limit = environment.get_value('FAIL_RETRIES')
 
-  for _ in range(retry_limit):
+  # TODO(mbarbella): Require callers to pass strings or bytes then reduce this
+  # back to the original retry limit.
+  for _ in range(retry_limit + 1):
     try:
       with open(file_path, file_mode) as file_handle:
-        file_handle.write(content_string)
-    except:
+        file_handle.write(content)
+    except TypeError:
+      content = str(content).encode('utf-8')
```

Comment:
Dont understand what this branch does.

Reply:
Added a few comments for clarity.

There are a few problems here, but all stem from the fact that we want this to accept arbitrary input. The two specific problems here are that in Python 3 write fails if the input isn't bytes if opened as wb (or str if opened w), and that we can no longer call encode() on a bytes to keep it as-is (it's no longer defined). This hack lets us catch the type errors and encode only if it was required, without breaking when we do pass bytes.

I think the real fix would be to expect a bytes as input to this function, but that would be a very large and invasive cleanup that can come later.